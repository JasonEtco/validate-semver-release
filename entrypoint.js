const fs = require('fs')
const path = require('path')
const semver = require('semver')
const { Toolkit } = require('actions-toolkit')
const core = require('@actions/core')

Toolkit.run(tools => {
  const pkg = tools.getPackageJSON()
  const { release } = tools.context.payload

  if (release.draft) {
    return tools.exit.neutral('This release is a draft! Aborting.')
  }

  const tag = semver.valid(release.tag_name)
  if (!tag) {
    return tools.exit.failure(`The tag ${release.tag_name} is not a valid tag.`)
  }

  if (pkg.version !== tag) {
    return tools.exit.failure(`Tag ${tag} and version in the package.json ${pkg.version} are not the same.`)
  }

  if (release.prerelease) {
    const prereleaseTag = semver.prerelease(release.tag_name)

    if (prereleaseTag === null) {
      return tools.exit.failure('The release is a prerelease, but the version tag is not.')
    }

    const VALID_TAGS = ['alpha', 'beta', 'next', 'rc']
    const [tagName] = prereleaseTag

    if (!VALID_TAGS.includes(tagName)) {
      return tools.exit.failure(`Publish tag ${tagName} is not a valid tag - it must be one of ${VALID_TAGS.join(', ')}`)
    }

    core.setOutput('release-workflow-tag', tagName)
    fs.writeFileSync(path.join(tools.workspace, 'release-workflow-tag'), tagName)
  }

  return tools.exit.success(`Release of version ${tag} is all set!`)
}, { event: 'release' })
