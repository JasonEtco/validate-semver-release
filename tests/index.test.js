const path = require('path')
const { Toolkit } = require('actions-toolkit')

function mockToolkit (event, workspace = 'workspace') {
  // Load the relevant JSON file
  process.env.GITHUB_EVENT_PATH = path.join(
    __dirname,
    'fixtures',
    `${event}.json`
  )
  // Load the relevant workspace file
  process.env.GITHUB_WORKSPACE = path.join(
    __dirname,
    'fixtures',
    workspace
  )

  // Silence warning
  Toolkit.prototype.warnForMissingEnvVars = jest.fn()

  // Mock the toolkit exit functions
  const tools = new Toolkit()
  tools.exit.neutral = jest.fn()
  tools.exit.failure = jest.fn()
  tools.exit.success = jest.fn()
  return tools
}

describe('validate-semver-release', () => {
  let runAction

  beforeEach(() => {
    Toolkit.run = jest.fn(fn => { runAction = fn })
    require('../entrypoint')

    Object.assign(process.env, {
      GITHUB_EVENT: 'release',
      GITHUB_WORKSPACE: path.join(__dirname, 'fixtures', 'workspace')
    })
  })

  it('exits neutral for draft releases', () => {
    const tools = mockToolkit('release-draft')
    runAction(tools)
    expect(tools.exit.neutral).toHaveBeenCalled()
    expect(tools.exit.neutral.mock.calls).toMatchSnapshot()
  })

  it('exits failure for invalid tag', () => {
    const tools = mockToolkit('release-invalid-tag')
    runAction(tools)
    expect(tools.exit.failure).toHaveBeenCalled()
    expect(tools.exit.failure.mock.calls).toMatchSnapshot()
  })

  it('exits failure for a tag/version mismatch', () => {
    const tools = mockToolkit('release-wrong-version')
    runAction(tools)
    expect(tools.exit.failure).toHaveBeenCalled()
    expect(tools.exit.failure.mock.calls).toMatchSnapshot()
  })

  it('exits successfully with no issues', () => {
    const tools = mockToolkit('release')
    runAction(tools)
    expect(tools.exit.success).toHaveBeenCalled()
    expect(tools.exit.success.mock.calls).toMatchSnapshot()
  })

  describe('prereleases', () => {
    it('exits failure with no semver prelease on the tag', () => {
      const tools = mockToolkit('release-prerelease-not-tag')
      runAction(tools)
      expect(tools.exit.failure).toHaveBeenCalled()
      expect(tools.exit.failure.mock.calls).toMatchSnapshot()
    })

    it('exits failure with an invalid prerelease tag', () => {
      const tools = mockToolkit('release-prerelease-invalid-tag', 'workspace-invalid-prerelease')
      runAction(tools)
      expect(tools.exit.failure).toHaveBeenCalled()
      expect(tools.exit.failure.mock.calls).toMatchSnapshot()
    })

    it('writes to a file with a valid tag and version', () => {
      const tools = mockToolkit('release-prerelease', 'workspace-prerelease')
      runAction(tools)
      expect(tools.exit.success).toHaveBeenCalled()
      expect(tools.exit.success.mock.calls).toMatchSnapshot()
      const file = tools.getFile('release-workflow-tag')
      expect(file).toBe('beta')
    })
  })
})
