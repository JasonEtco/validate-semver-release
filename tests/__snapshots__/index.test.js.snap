// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate-semver-release exits failure for a tag/version mismatch 1`] = `
Array [
  Array [
    "Tag 2.0.0 and version in the package.json 1.0.0 are not the same.",
  ],
]
`;

exports[`validate-semver-release exits failure for invalid tag 1`] = `
Array [
  Array [
    "The tag pizzadog is not a valid tag.",
  ],
]
`;

exports[`validate-semver-release exits neutral for draft releases 1`] = `
Array [
  Array [
    "This release is a draft! Aborting.",
  ],
]
`;

exports[`validate-semver-release exits successfully with no issues 1`] = `
Array [
  Array [
    "Release of version 1.0.0 is all set!",
  ],
]
`;

exports[`validate-semver-release prereleases exits failure with an invalid prerelease tag 1`] = `
Array [
  Array [
    "Publish tag meow is not a valid tag - it must be one of beta, next",
  ],
]
`;

exports[`validate-semver-release prereleases exits failure with no semver prelease on the tag 1`] = `
Array [
  Array [
    "The release is a prerelease, but the version tag is not.",
  ],
]
`;

exports[`validate-semver-release prereleases writes to a file with a valid tag and version 1`] = `
Array [
  Array [
    "Release of version 1.0.0-beta.1 is all set!",
  ],
]
`;
